protocol-specs/generators/dpos_random_seed_generation/index.js

```diff
@@ -0,0 +1,433 @@
+		const randomSeed1 = H(m + 1);
+		const randomSeed2 = H(0);
+
+		return { randomSeed1, randomSeed2 };
+	}
+
+	const { firstBlock: firstBlockBlockOfLastBlockRound } = roundFlags(
+		r.round - 1,
+		n,
+	);
+
+	const filteredBlocks = blocks.reduce((acc, b) => {
+		if (b.height >= firstBlockBlockOfLastBlockRound) {
+			acc[b.height] = b;
+		}
+		return acc;
+	}, {});
+
+	const seedRevealMap = Object.values(filteredBlocks).reduce((acc, b) => {
+		acc[b.height] = b.seedReveal;
+		return acc;
+	}, {});
+
+	const forgedBlocksMap = Object.values(filteredBlocks).reduce((acc, b) => {
+		if (!acc[b.generatorPublicKey]) {
+			acc[b.generatorPublicKey] = [];
+		}
+		acc[b.generatorPublicKey].push(b.height);
+		return acc;
+	}, {});
+
+	const seedRevealsForRandomSeed1 = [];
+
+	// From middle of current round to middle of last round
+	for (let i = h + m; i >= h - m; i -= 1) {
+		const { generatorPublicKey, seedReveal } = filteredBlocks[i];
+		const [previousForgedHeight] = forgedBlocksMap[generatorPublicKey];
+
+		// If forged only 1 block in current and last round
+		// don't use that seed reveal
+		if (!previousForgedHeight || previousForgedHeight === i) {
+			continue;
+		}
+
+		const previousSeedReveal = seedRevealMap[previousForgedHeight];
+
+		if (H(seedReveal) !== previousSeedReveal) {
+			continue;
+		}
+
+		seedRevealsForRandomSeed1.push(seedReveal);
+	}
+
+	const seedRevealsForRandomSeed2 = [];
+	// From middle of current round to middle of last round
+	for (let i = h - 1; i >= h - n; i -= 1) {
+		const { generatorPublicKey, seedReveal } = filteredBlocks[i];
+		const [previousForgedHeight] = forgedBlocksMap[generatorPublicKey];
+
+		// If forged only 1 block in current and last round
+		// don't use that seed reveal
+		if (!previousForgedHeight || previousForgedHeight === i) {
+			continue;
+		}
+
+		const previousSeedReveal = seedRevealMap[previousForgedHeight];
+
+		if (H(seedReveal) !== previousSeedReveal) {
+			continue;
+		}
+
+		seedRevealsForRandomSeed1.push(seedReveal);
+	}
+
+	const randomSeed1 = XOR([H(h + m), ...seedRevealsForRandomSeed1]);
+	const randomSeed2 = XOR([H(h - 1), ...seedRevealsForRandomSeed2]);
+
+	return { randomSeed1, randomSeed2 };
+};
+
+const randomSeedFirstRound = () => ({
+	title: 'Random seed for first round',
+	summary: 'Random seeds generation for first round',
+	config: 'devnet',
+	runner: 'dpos_random_seed_generation',
+	handler: 'dpos_random_seed_generation_first_round',
+	testCases: (() => {
+		const blocksPerRound = activeDelegates + standByDelegates;
+		const blocks = generateBlocks({
+			startHeight: 1,
+			numberOfBlocks: blocksPerRound,
+			delegateList: sampleDelegateList.slice(0, blocksPerRound),
+		});
+		const { randomSeed1, randomSeed2 } = generateRandomSeed(
+			blocks,
+			blocksPerRound,
+		);
+
+		return [
+			{
+				input: {
+					blocksPerRound,
+					blocks,
+				},
+				output: {
+					randomSeed1,
+					randomSeed2,
+				},
+			},
+		];
+	})(),
+});
+
+const randomSeedSecondRound = () => ({
```

Comment:
I would change this case so it covers the general case (when everything is ok). So it should be at a generic round r (when r > 3) and considering the corresponding seedReveal values in the past to   check if they are valid in generate the next random seeds.